sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -657,18 +656,41 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
 
     val useAdvanced = SQLConf.get.advancedPartitionPredicatePushdownEnabled
 
+    object ExtractAttribute {
+      def unapply(expr: Expression): Option[Attribute] = {
+        expr match {
+          case attr: Attribute => Some(attr)
+          case cast @ Cast(child, dt, _) if !Cast.mayTruncate(child.dataType, dt) => unapply(child)
+          case _ => None
+        }
+      }
+    }
+
+    def decoratedValue(attr: Attribute, value: String): String = {
```

Comment:
what does it do? can you add some comments?

Reply:
e.g. 
dt="20180101"  if dt is of NumericType, we remove the quote and it changes to be  dt=20180101
dt=201880101 if dt is of StringType, we add the quote and it changes to be dt="20180101"